bumble/device.py

```diff
@@ -1141,17 +1157,121 @@ async def on_inquiry_complete(self):
 
     @host_event_handler
     @with_connection_from_handle
-    def on_connection_authentication_complete(self, connection):
-        logger.debug(f'*** Connection Authentication Complete: [0x{connection.handle:04X}] {connection.peer_address} as {connection.role_name}')
+    def on_connection_authentication(self, connection):
+        logger.debug(f'*** Connection Authentication: [0x{connection.handle:04X}] {connection.peer_address} as {connection.role_name}')
         connection.authenticated = True
-        connection.emit('connection_authentication_complete')
+        connection.emit('connection_authentication')
 
     @host_event_handler
     @with_connection_from_handle
     def on_connection_authentication_failure(self, connection, error):
         logger.debug(f'*** Connection Authentication Failure: [0x{connection.handle:04X}] {connection.peer_address} as {connection.role_name}, error={error}')
         connection.emit('connection_authentication_failure', error)
 
+    # [Classic only]
+    @host_event_handler
+    @with_connection_from_address
+    def on_authentication_io_capability_request(self, connection):
+        # Ask what the pairing config should be for this connection
+        pairing_config = self.pairing_config_factory(connection)
+
+        # Map the SMP IO capability to a Classic IO capability
```

Comment:
Shouldn’t we have an additional abstraction on top of pair/authenticate?

* We have a `pair` method for BLE only (it may be worth adding a comment on it btw) and `authenticate` for classic only, which is a bit confusing from a user perspective.
* The current structure uses SMP global variables within classic pairing methods (for IO caps), which should be fine for most users (since it is abstracted in `PairingDelegate`), but a bit confusing for a Bluetooth informed developer (since SMP is supposed to handle BLE only).
* It may simplify implementing Cross Transport Key Derivation (CTKD) if we are able to share keys as well. 

Ideally, I would like to define IO caps for my device (not for a connection), call `pair` (or `authenticate`) on a specific connection (with BR/EDR or BLE transport), and, if there is another connection with the same peer using another transport, be able to use CTKD (if enabled) to avoid having to redo the pairing process.

Reply:
Yes, we should definitely try to unify the API for Classic `authenticate` and BLE `pair`. I was leaving that as a future enhancement, once the underlying transport-specific APIs are well implemented.
The general idea is that most developers are going to be interested in either BLE or Classic, and only a small subset will be interested in both at the same time. The goal is to offer common abstractions when they come at no cost (or close to no cost) to the transport-specific users, so that we don't burden single-transport users with things that aren't relevant to them.
About key sharing, that should already be supported, since the notion of key store is the same regardless of the transport. But dual-connections haven't really been tested yet, and thus CTKD is still on the TODO list.

I also agree that it would be nice to refactor the code a bit in the future to move the PairingDelegate and PairingConfig in a transport-independent place, and update the SMP implementation to use that transport-independent class with mappings back to the BLE-specific values. That can be done in a future PR (the fact that the smp classes are used internally here isn't exposed to the users of the `Device` class, so that sort of refactoring should be fairly easy, just an import statement change.

Reply:
Sounds good, let's keep that as a future enhancement 👍